File: test/test_kruX.py
import numpy
import subprocess
import time
import os
import sys
sys.path.append('../python/')
import kruX

def test():
  """ Test the results from R's build-in kruskal wallis function, R's matrix-based
  kw test, and Python's matrix-based kw test
  directory.  Can be done in a nice way with rpy
  """
  subprocess.call(['Rscript','test_kruX.R'])
  outputFromR = numpy.genfromtxt('R.output',
                dtype=[('SNP','i4'),('gene','i4'),('fd','i4'),
                       ('chi','f4'),('pvalue','f4') ] )
  geno = numpy.genfromtxt('geno.tab.tmp', skip_header=1, missing_values='NA')
  geno = geno[:,1:]
  mrna = numpy.genfromtxt('mrna.tab.tmp', skip_header=1, missing_values='NA')
  mrna = mrna[:,1:] 
  start = time.clock()
  outputFromPython = kruX.calculateKruskalWallisWithMatrix(geno, mrna, 
                     pValueThre = 1, numTranscript=1000)
  endTime = time.clock() - start
  print('Total running time for Python matrix-based test: %.2f'%(endTime))
  numpy.savetxt('python.output', outputFromPython, fmt='%i\t%i\t%i\t%.17e\t%.17e',delimiter='\t')
  if (numpy.all(outputFromR['pvalue']-outputFromPython['pvalue']<1e-7)):
    print('Passed: equal between R matrix-based  and Python matrix-based test')
  else:  
    print('Failed: not equal between R matrix-based  and Python matrix-based test')

def test_python_performance():
  """Python's matrix-based kw test running time
  """
  geno = numpy.genfromtxt('geno.tab.tmp', skip_header=1, missing_values='NA')
  geno = geno[:,1:]
  mrna = numpy.genfromtxt('mrna.tab.tmp', skip_header=1, missing_values='NA')
  mrna = mrna[:,1:]
  start = time.clock()
  outputFromPython = kruX.calculateKruskalWallisWithMatrix(geno, mrna,
                     pValueThre = 1)
  endTime = time.clock() - start
  print('Total running time for Python matrix-based test: %.2f'%(endTime))
  numpy.savetxt('python.output', outputFromPython, fmt='%i\t%i\t%i\t%.7f\t%.7f',delimiter='\t')

if __name__ == '__main__':
  test()
  #test_python_performance()

File: test/test_kruX.R
##Test script between R built-in function and matrix-based function
source('../kruX.R')
args <- commandArgs(trailingOnly = TRUE)
num.sample=as.integer(args[1])
num.snp=as.integer(args[2])
num.mrna=as.integer(args[3])
per.missing = as.numeric(args[4])
p.value.threshold = as.numeric(args[5])

cat(args[1], ' samples \n', 
    args[2],     ' SNPs    \n',
    args[3],    ' transcripts \n' ,
    args[4] , ' missing values \n' ,
    args[5] , ' pvalue threshold \n')
geno=matrix(sample(c(0,1,2, NA), size=num.sample*num.snp, 
                  replace=T, prob=c(0.32,0.32,0.32,0.04)),
                 nrow=num.snp, ncol=num.sample,
                 dimnames=list(seq(0,num.snp-1), seq(0,num.sample-1))) 

mrna=matrix(rnorm(num.mrna*num.sample),
                  nrow=num.mrna, ncol=num.sample,
                  dimnames=list(seq(0,num.mrna-1), seq(0,num.sample-1)))
mrna[cbind(sample(dim(mrna)[1],round(num.mrna*num.sample*per.missing), replace=T), 
     sample(dim(mrna)[2],round(num.mrna*num.sample*per.missing), replace=T))] <- NA

write.table(geno, file='geno.tab.tmp' ,sep='\t', quote=F)
write.table(mrna, file='mrna.tab.tmp' ,sep='\t', quote=F)

t.start = proc.time()
res1=CalculateKruskalWallisWithMatrix(geno, mrna, p.value.threshold)
res1=do.call(rbind, res1)
cat("Total running time for R matrix-based test: ", proc.time()[3]-t.start[3],"\n")
write.table(res1[order(res1$pvalue),], file='R.output', quote=F, row.names=F, col.names=F)
res1=res1[order(res1$pvalue),]

res2=brutal.force.kruskal(geno, mrna, p.value.threshold)
res2=res2[order(res2$pvalue),]

if (T %in% (res1$pvalue-res2$pvalue)>1e-10){
  cat('failed: R built-in function vs R matrix-based test \n')
} else{
  cat('passed: R built-in function vs R matrix-based test  \n')
}
